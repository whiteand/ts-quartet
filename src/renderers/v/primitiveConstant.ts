import { TPrimitiveSchema } from "../../types";
import { getDispatchedRenderer, getNotImplementedRenderer } from "../utils";
import { booleanConstantRenderer } from "./booleanConstant";
import { nullConstantRenderer } from "./nullConstant";
import { numberConstantRenderer } from "./numberConstant";
import { stringConstantRenderer } from "./stringConstant";
import { symbolConstantRenderer } from "./symbolConstant";
import { undefinedConstantRenderer } from "./undefinedConstant";

type PrimitiveWithoutNullOrUndefinedTypes =
  | "boolean"
  | "symbol"
  | "string"
  | "number"
  | "bigint";

const RENDERER_FOR_TYPE = {
  bigint: getNotImplementedRenderer(
    new TypeError(
      'bigint is not supported by quartet. Use v.custom(value => typeof value === "bigint")'
    )
  ),
  boolean: booleanConstantRenderer,
  number: numberConstantRenderer,
  string: stringConstantRenderer,
  symbol: symbolConstantRenderer
};

export const primitiveConstantRenderer = getDispatchedRenderer<
  TPrimitiveSchema
>(schema => {
  if (schema === null) {
    return nullConstantRenderer;
  }
  if (schema === undefined) {
    return undefinedConstantRenderer;
  }
  const type = typeof schema as PrimitiveWithoutNullOrUndefinedTypes;
  return RENDERER_FOR_TYPE[type];
});
