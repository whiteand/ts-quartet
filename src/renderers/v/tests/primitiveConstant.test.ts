import { primitiveConstantRenderer } from "../primitiveConstant";
import { snapshot } from "./snapshot";

describe("primitiveConstantRenderer", () => {
  test("v(null)", () => {
    snapshot(primitiveConstantRenderer, { schema: null });
  });
  test("v(undefined)", () => {
    snapshot(primitiveConstantRenderer, { schema: undefined });
  });
  test("v(true)", () => {
    snapshot(primitiveConstantRenderer, { schema: true });
  });
  test("v(false)", () => {
    snapshot(primitiveConstantRenderer, { schema: false });
  });
  test("v(symbol)", () => {
    snapshot(primitiveConstantRenderer, { schema: Symbol.for("quartet") });
  });
  test("v(string)", () => {
    snapshot(primitiveConstantRenderer, { schema: "Andrew" });
  });
  test("v(long string)", () => {
    snapshot(primitiveConstantRenderer, {
      schema:
        "Very very veryVery very veryVery very veryVery very veryVery very very long list"
    });
  });
  test("v(NaN)", () => {
    snapshot(primitiveConstantRenderer, { schema: 0 / 0 });
  });
  test("v(number)", () => {
    snapshot(primitiveConstantRenderer, { schema: 42 });
  });
});
