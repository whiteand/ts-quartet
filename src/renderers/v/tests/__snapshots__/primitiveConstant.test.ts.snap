// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`primitiveConstantRenderer v(NaN) 1`] = `"Number.isNaN(_valueId)"`;

exports[`primitiveConstantRenderer v(NaN) 2`] = `"!Number.isNaN(_valueId)"`;

exports[`primitiveConstantRenderer v(NaN) 3`] = `"if (Number.isNaN(_valueId)) return true;"`;

exports[`primitiveConstantRenderer v(NaN) 4`] = `"if (!Number.isNaN(_valueId)) return false;"`;

exports[`primitiveConstantRenderer v(false) 1`] = `"_valueId === false"`;

exports[`primitiveConstantRenderer v(false) 2`] = `"_valueId !== false"`;

exports[`primitiveConstantRenderer v(false) 3`] = `"if (_valueId === false) return true;"`;

exports[`primitiveConstantRenderer v(false) 4`] = `"if (_valueId !== false) return false;"`;

exports[`primitiveConstantRenderer v(long string) 1`] = `"_valueId === _validString_\\"Very very veryVery very veryVery very veryVery very veryVery very very long list\\""`;

exports[`primitiveConstantRenderer v(long string) 2`] = `"_valueId !== _validString_\\"Very very veryVery very veryVery very veryVery very veryVery very very long list\\""`;

exports[`primitiveConstantRenderer v(long string) 3`] = `"if (_valueId === _validString_\\"Very very veryVery very veryVery very veryVery very veryVery very very long list\\") return true;"`;

exports[`primitiveConstantRenderer v(long string) 4`] = `"if (_valueId !== _validString_\\"Very very veryVery very veryVery very veryVery very veryVery very very long list\\") return false;"`;

exports[`primitiveConstantRenderer v(null) 1`] = `"_valueId === null"`;

exports[`primitiveConstantRenderer v(null) 2`] = `"_valueId !== null"`;

exports[`primitiveConstantRenderer v(null) 3`] = `"if (_valueId === null) return true;"`;

exports[`primitiveConstantRenderer v(null) 4`] = `"if (_valueId !== null) return false;"`;

exports[`primitiveConstantRenderer v(number) 1`] = `"_valueId === _validNumber_42"`;

exports[`primitiveConstantRenderer v(number) 2`] = `"_valueId !== _validNumber_42"`;

exports[`primitiveConstantRenderer v(number) 3`] = `"if (_valueId === _validNumber_42) return true;"`;

exports[`primitiveConstantRenderer v(number) 4`] = `"if (_valueId !== _validNumber_42) return false;"`;

exports[`primitiveConstantRenderer v(string) 1`] = `"_valueId === \\"Andrew\\""`;

exports[`primitiveConstantRenderer v(string) 2`] = `"_valueId !== \\"Andrew\\""`;

exports[`primitiveConstantRenderer v(string) 3`] = `"if (_valueId === \\"Andrew\\") return true;"`;

exports[`primitiveConstantRenderer v(string) 4`] = `"if (_valueId !== \\"Andrew\\") return false;"`;

exports[`primitiveConstantRenderer v(symbol) 1`] = `"_valueId === _validSymbol_undefined"`;

exports[`primitiveConstantRenderer v(symbol) 2`] = `"_valueId !== _validSymbol_undefined"`;

exports[`primitiveConstantRenderer v(symbol) 3`] = `"if (_valueId === _validSymbol_undefined) return true;"`;

exports[`primitiveConstantRenderer v(symbol) 4`] = `"if (_valueId !== _validSymbol_undefined) return false;"`;

exports[`primitiveConstantRenderer v(true) 1`] = `"_valueId === true"`;

exports[`primitiveConstantRenderer v(true) 2`] = `"_valueId !== true"`;

exports[`primitiveConstantRenderer v(true) 3`] = `"if (_valueId === true) return true;"`;

exports[`primitiveConstantRenderer v(true) 4`] = `"if (_valueId !== true) return false;"`;

exports[`primitiveConstantRenderer v(undefined) 1`] = `"_valueId === undefined"`;

exports[`primitiveConstantRenderer v(undefined) 2`] = `"_valueId !== undefined"`;

exports[`primitiveConstantRenderer v(undefined) 3`] = `"if (_valueId === undefined) return true;"`;

exports[`primitiveConstantRenderer v(undefined) 4`] = `"if (_valueId !== undefined) return false;"`;
